Add BroadcastDistributionRule

jihoonson opened the pull request:
First patch for #4032.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/druid-io/druid/4077)
<!-- Reviewable:end -->

jon-wei commented on server/src/main/java/io/druid/server/coordinator/rules/BroadcastDistributionRule.java:
```diff
@@ -0,0 +1,90 @@
+            "Failed to replicate segment[%s] to server[%s] due to insufficient available space",
+            segment.getIdentifier(),
+            holder.getServer().getHost()
+        )
```
is the formatting off here?

jihoonson replied:
No.. This is druid's format if i didn't make a mistake. Maybe better if i use a separate string value for the error message like below?
```java
String message = String.format(...);
log.makeAlert(message)
    .addData("segmentSize", segment.getSize())
    .addData("availableSize", holder.getAvailableSize())
    .emit();
```

jon-wei commented on server/src/main/java/io/druid/server/coordinator/rules/ForeverBroadcastDistributionRule.java:
```diff
@@ -0,0 +1,84 @@
+{
+  static final String TYPE = "broadcastForever";
+
+  private final String colocateDataSource;
```
suggest renaming this here and elsewhere to `colocatedDatasource`

jihoonson replied:
Thanks. Done.

weijietong commented on server/src/main/java/io/druid/server/coordinator/rules/BroadcastDistributionRule.java:
```diff
@@ -0,0 +1,90 @@
+           .addData("availableSize", holder.getAvailableSize())
+           .emit();
+      } else {
+        holder.getPeon().loadSegment(
```
here not found replicator support , or next submit ?

jihoonson replied:
You mean ```ReplicationThrottler```? 
```ReplicationThrottler``` is used to throttle replication speed, and I think it's not necessary for broadcast rules because users usually want that broadcast completes as soon as possible.

gianm commented on server/src/main/java/io/druid/server/coordinator/rules/ForeverBroadcastDistributionRule.java:
```diff
@@ -0,0 +1,83 @@
+      @JsonProperty("colocatedDatasource") String colocatedDatasource
+  )
+  {
+    this.colocatedDatasource = Objects.requireNonNull(colocatedDatasource);
```
Most other Druid code uses `Preconditions.checkNotNull` but I guess there's no real need to be consistent there.

gianm commented on server/src/main/java/io/druid/server/coordinator/rules/BroadcastDistributionRule.java:
```diff
@@ -0,0 +1,89 @@
+    // Find servers which holds the segments of co-located data source
+    final List<ServerHolder> targetServerHolders = params
+        .getDruidCluster().getAllServers().stream()
+        .filter(eachHolder -> eachHolder.getServer().getDataSource(getColocatedDatasource()) != null)
```
It seems to me like it'd be valuable to have `colocatedDataSources` be a list, and maybe if it's empty then broadcast to _all_ servers. That'd give some more flexibilty to the user about where to broadcast their data sources.

jihoonson replied:
Thanks. Sounds nice.

gianm commented on server/src/main/java/io/druid/server/coordinator/rules/BroadcastDistributionRule.java:
```diff
@@ -0,0 +1,89 @@
+    stats.addToGlobalStat(LoadRule.ASSIGNED_COUNT, 0);
+
+    for (ServerHolder holder : serverHolderList) {
+      if (segment.getSize() > holder.getAvailableSize()) {
```
Some sites use the main alert text to collect and summarize similar alerts. So it'd be nicer for this description to be `("Failed to broadcast segment for[%s]", dataSource)` with the segment id and specific server as additional data.

gianm commented on server/src/main/java/io/druid/server/coordinator/rules/BroadcastDistributionRule.java:
```diff
@@ -0,0 +1,89 @@
+    return stats;
+  }
+
+  public abstract String getColocatedDatasource();
```
Please spell this as getColocatedDataSource (or dataSources) not datasource.

gianm commented on server/src/main/java/io/druid/server/coordinator/rules/ForeverBroadcastDistributionRule.java:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  @Override
+  public boolean equals(Object o)
```
Should have hashCode overridden too so it is consistent with equals.

gianm replied:
Please override toString too, I think there's some code that prints rules.

jihoonson replied:
Would you give me some examples where ```toString``` is used? It's not overridden for other rules as well.

gianm replied:
Ah, nevermind, you're right.

gianm commented on server/src/main/java/io/druid/server/coordinator/rules/IntervalBroadcastDistributionRule.java:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  @Override
+  public boolean equals(Object o)
```
Override hashCode too

weijietong commented:
@jihoonson  I have read the code of this part PR . The 'BroadcastDistributionRule' is a really good way to express DataSource distribution for the join target . 

But  another thing needs to be care about is `DruidCoordinatorBalancer`. As I ever mentioned , Druid now hard code using `CostBalancerStrategy` . When the `DruidCoordinatorBalancer`  runs periodically, it may breaks the `DataSource` distribution specified by the broadcast or LocalJoin load rule. 

My solution is to rewrite the `CostBalancerStrategy` to use my specific BalancerStrategy first to pick a segment to move . Then check the segment DataSource again to enforce it's movement not to break the distribution rule.  A little indirect but not extensible . I wonder whether you have thought about this and anyone else has a good solution ?

jihoonson commented:
@gianm thanks. Addressed your comments. I'll document soon.
@weijietong I haven't thought about that yet, but your idea sounds to make sense to me. I'll think about that too.

jihoonson commented:
The travis failure seems not related to this patch. @jon-wei, would you restart the build? I don't have permission.

jihoonson commented:
I updated the patch to drop segments which are not co-located anymore.

fjy commented on server/src/main/java/io/druid/server/coordinator/rules/BroadcastDistributionRule.java:
```diff
@@ -0,0 +1,112 @@
+
+    final CoordinatorStats stats = new CoordinatorStats();
+
+    return stats.accumulate(assign(loadServerHolders, segment))
```
im wondering if the broadcast rules should label their segments a little differently, or at least be able to distinguish from actual Druid segments

fjy replied:
It'll definitely make reading the logs easier

jihoonson replied:
I think broadcasted segments should be regarded as same as other segments. Every types of queries can be queried on broadcasted segments, but the broadcast information is required by only joins.

The information of which sources are broadcasted or not is required by brokers and historicals (and realtimes in the future). Brokers decides a given query should be distributed to which nodes, and this is determined by selecting the nodes holding the segments of non-broadcasted tables. Historicals, as proposed in #4032, join broadcasted segments first and then subsequently join the result with non-broadcasted segments in parallel.

I'm currently thinking that brokers are able to figure out which tables are broadcasted by looking at BroadcastDistributionRules, and they will add this information to QueryContext for historicals. If this works, broadcast segments can be regarded as normal. 

What do you think?

fjy replied:
@jihoonson Okay that makes sense to me. Can broadcast segments be created dynamically?

jihoonson replied:
Maybe possible in the future when we support the feature of broadcasting tables on the fly, but not yet.

jihoonson commented:
I found two problems when this patch is applied. 
1) Segments of broadcast data sources might not be fully co-located with segments of colocatedDataSources if load/drop rules for colocatedDataSources are run before BroadcastDistributionRule.
2) Since DruidCoordinatorBalancer is run after DruidCoordinatorRuleRunner, DruidCoordinatorBalancer breaks the co-location of broadcast segments and colocatedDataSources' segments.

The solution I'm currently considering is as follows.
1) Building a dependency graph of rules, and run rules from independent ones to dependent ones. There must be no cycle in the dependency graph. This solution changes Druid's rule application policy, '[_first matched rule is applied_](http://druid.io/docs/latest/operations/rule-configuration.html)', so we need more discussion.
2) Changing the unit for load balancing from a segment to a set of co-located segments. The cost for moving co-located segments will be definitely large, so usually independent segments will be chosen for load balancing. 

Since BroadcastDistributionRules are just introduced here, and not used in anywhere, I think it would be ok to fix these problems in follow-up PRs.

fjy commented:
👍

gianm commented:
@jihoonson, on the two points you raised, it sounds like the issue is that (a) loading new segments, and (b) moving existing segments for balancing, could break broadcasting if a segment from the colocated datasource is moved onto a historical node that didn't already have _any_ segments from that datasource.

I guess in practice this is potentially an issue, but won't be a huge one, since odds are that a distributed datasource will have at least one segment on all historical nodes in a given tier anyway (unless it has a number of segments smaller than the number of historical nodes).

To keep things (very) simple, what do you think about removing the "colocatedDataSources" concept, and instead broadcasting the broadcast datasources to every historical node the given tier for the rule? I think in practice this achieves mostly the same thing (given the reason above) as having colocatedDataSources. It also avoids bugs due to this effect, since every historical in a tier will have the broadcast segments and there's no need to worry.

gianm commented:
Or at least, making that the behavior if colocatedDataSources is empty, so users can leave it empty and not have to worry.

gianm commented:
The docs should mention this caveat.

jihoonson commented:
@gianm sounds good, but I'm concerned with the usability. If we support broadcasting to tiers instead of broadcasting to data sources, users will need two steps to design their data distribution. That is, they first need to check each data source to be joined is involved which tiers, and then decide which data sources need to be broadcasted to which tiers. Broadcasting to data sources will be simpler because users need to decide which sources are broadcasted to which source.
In addition, if load rules for a data source are changed, broadcast rules relating to that data source should be changed accordingly. This will be inconvenient.
What do you think?

jihoonson commented:
@gianm if colocatedDataSources is null or empty, segments are broadcasted to all nodes.

jihoonson commented:
Maybe we can support two different types of broadcast rules for 1) broadcasting to tiers and 2) colocating data sources.

gianm commented:
The latter suggestion sounds good. I think in practice many users will choose to broadcast to tiers -- since if you're already very aware of your tier structure (as many users are, if they use tiers) then it should be pretty easy to choose the right tier for a broadcast. It also works around the issue you brought up, which is nice.

The colocated data source option is a nice alternative that would make configuration simpler, just should come with a caveat about the issue you brought up.

jihoonson commented:
Ok, then I'll make another PR for broadcasting to tiers.

jihoonson commented:
@gianm, I added a caveat

gianm commented on docs/content/operations/rule-configuration.md:
```diff
@@ -126,6 +129,65 @@ Period drop rules are of the form:
+The interval of a segment will be compared against the specified period. The period is from some time in the past to the current time. The rule matches if the period contains the interval.
+
+<div class="note caution">
+broadcastToDataSources rules don't guarantee that segments of the data sources are always co-located.
```
This comment would be better if it explained why. Maybe adding "because segments for the colocated data sources are not loaded together atomically".

jihoonson replied:
Added. Thanks.

gianm commented on docs/content/operations/rule-configuration.md:
````diff
@@ -126,6 +129,65 @@ Period drop rules are of the form:
+
+```json
+{
+  "type" : "broadcastToDataSourcesForever",
````
I think this should just be called `broadcastForever`, since the `dataSources` part is optional. And similar comment for the other kinds of broadcast rules.

jihoonson replied:
Renamed.

gianm commented on server/src/main/java/io/druid/server/coordinator/rules/BroadcastDistributionRule.java:
```diff
@@ -0,0 +1,112 @@
+    stats.addToGlobalStat(LoadRule.ASSIGNED_COUNT, 0);
+
+    for (ServerHolder holder : serverHolders) {
+      if (segment.getSize() > holder.getAvailableSize()) {
```
I think that like CostBalancerStrategy, this should check if `holder.isLoadingSegment(segment)` and skip if so.

jihoonson replied:
Thanks! I added it.

gianm commented on server/src/main/java/io/druid/server/coordinator/rules/BroadcastDistributionRule.java:
```diff
@@ -0,0 +1,112 @@
+            if (colocatedDataSources.stream()
+                                    .anyMatch(source -> eachHolder.getServer().getDataSource(source) != null)) {
+              loadServerHolders.add(eachHolder);
+            } else if (eachHolder.isServingSegment(segment)) {
```
I think this should skip holders that already have the segment in their drop lists (`peon.getSegmentsToDrop()`), like DruidCoordinator.moveSegment and DruidCoordinatorCleanupUnneeded.

jihoonson replied:
Thanks! I added it.

jihoonson commented on server/src/main/java/io/druid/server/coordinator/DruidCluster.java:
```diff
@@ -70,6 +72,15 @@ public void add(ServerHolder serverHolder)
 
+  public Set<ServerHolder> getAllServers()
+  {
+    final Set<ServerHolder> servers = new HashSet<>();
```
I misunderstood that the same server can be involved in different tiers which it can't. I changed to list.

jihoonson replied:
@gianm I found ```ServerHolder.equals()``` calls ```ImmutableDruidServer.equals()``` and ```LoadQueuePeon.equals()``` which are not implemented. Maybe this is not a critical problem for now because it is not called so far.. Anyway, I'll fix it. 
Btw, to check the equality of two ```ServerHolder```s, I think it would be enough to compare ImmutableDruidServer.getHost(), ImmutableDruidServer.getType(), and ImmutableDruidServer.getTier(). What do you think?
